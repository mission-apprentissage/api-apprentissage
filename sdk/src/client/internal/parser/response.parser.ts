import { z } from "zod/v4-mini";
import { $ZodError, treeifyError } from "zod/v4/core";
import type { $ZodType } from "zod/v4/core";

export class ApiParseError extends Error {
  constructor(zodError: $ZodError) {
    super(
      "api-alternance-sdk: Error while parsing API response, this is either a bug in the SDK or a change in the API response. Please report this issue.\n" +
        JSON.stringify(treeifyError(zodError))
    );
    this.name = "ApiParseError";
    this.cause = zodError;
  }
}

export function parseApiResponse<T extends $ZodType>(data: unknown, schema: T): z.output<T> {
  try {
    return z.parse(schema, data);
  } catch (error) {
    if (error instanceof $ZodError) {
      throw new ApiParseError(error);
    }
    throw error;
  }
}
