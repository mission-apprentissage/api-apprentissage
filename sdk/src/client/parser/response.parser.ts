import type { z, ZodType } from "zod";
import { ZodError } from "zod";

export class ApiParseError extends Error {
  constructor(zodError: ZodError) {
    super(
      "api-alternance-sdk: Error while parsing API response, this is either a bug in the SDK or a change in the API response. Please report this issue.\n\tZodErrors: " +
        zodError.toString().split("\n").join("\n\t")
    );
    this.name = "ApiParseError";
    this.cause = zodError;
  }
}

export function parseApiResponse<T extends ZodType>(data: unknown, schema: T): z.output<T> {
  try {
    return schema.parse(data);
  } catch (error) {
    if (error instanceof ZodError) {
      throw new ApiParseError(error);
    }
    throw error;
  }
}
