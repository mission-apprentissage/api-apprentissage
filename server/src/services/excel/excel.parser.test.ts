import { createReadStream } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";
import { describe, expect, it } from "vitest";

import { parseExcelFileStream } from "./excel.parser.js";

async function getRows<T>(generator: AsyncGenerator<T>): Promise<T[]> {
  const data: T[] = [];
  for await (const row of generator) {
    data.push(row);
  }

  return data;
}

describe("parseExcelFileStream", () => {
  it("should parse the excel file stream", async () => {
    const dataFixture = join(dirname(fileURLToPath(import.meta.url)), "fixtures/file.xlsx");
    const s = createReadStream(dataFixture);

    const spec = [
      {
        type: "required",
        nameMatchers: [/Onglet 3 - référentiel NPEC/i],
        key: "npec",
        skipRows: 3,
        columns: [
          { type: "required", name: "rncp", regex: [/^Code RNCP$/i] },
          { type: "required", name: "formation_libelle", regex: [/^Libellé de la formation$/i] },
          { type: "required", name: "certificateur", regex: [/^Certificateur\*\s+/i] },
          { type: "required", name: "diplome_libelle", regex: [/^Libellé du Diplôme$/i] },
          { type: "required", name: "cpne_code", regex: [/^Code CPNE$/i] },
          { type: "required", name: "cpne_libelle", regex: [/^CPNE$/i] },
          { type: "required", name: "npec", regex: [/^NPEC final$/i] },
        ],
      },
      {
        type: "required",
        nameMatchers: [/Onglet 4 - CPNE-IDCC/i],
        key: "cpne-idcc",
        skipRows: 1,
        columns: [
          { type: "required", name: "cpne_code", regex: [/^Code CPNE/i] },
          { type: "required", name: "cpne_libelle", regex: [/^CPNE/i] },
          { type: "required", name: "idcc", regex: [/^IDCC/i] },
        ],
      },
    ] as const;

    const data = await getRows(parseExcelFileStream(s, spec));

    expect(data).toMatchSnapshot();
  });

  it("should skip parsing if spec is null", async () => {
    const dataFixture = join(dirname(fileURLToPath(import.meta.url)), "fixtures/file.xlsx");
    const s = createReadStream(dataFixture);

    const spec = [
      {
        type: "ignore",
        nameMatchers: [/Onglet 3 - référentiel NPEC/i],
      },
      {
        type: "required",
        nameMatchers: [/Onglet 4 - CPNE-IDCC/i],
        key: "cpne-idcc",
        skipRows: 1,
        columns: [
          { type: "required", name: "cpne_code", regex: [/^Code CPNE/i] },
          { type: "required", name: "cpne_libelle", regex: [/^CPNE/i] },
          { type: "required", name: "idcc", regex: [/^IDCC/i] },
        ],
      },
    ] as const;

    const data = await getRows(parseExcelFileStream(s, spec));

    expect(data).toMatchSnapshot();
  });

  it("should throw if sheet is not found", async () => {
    const dataFixture = join(dirname(fileURLToPath(import.meta.url)), "fixtures/file.xlsx");
    const s = createReadStream(dataFixture);

    const spec = [
      {
        type: "ignore",
        nameMatchers: [/Onglet 3 - référentiel NPEC/i],
      },
      {
        type: "required",
        nameMatchers: [/Onglet 4 - CPNE-IDCC/i],
        key: "cpne-idcc",
        skipRows: 1,
        columns: [
          { type: "required", name: "cpne_code", regex: [/^Code CPNE/i] },
          { type: "required", name: "cpne_libelle", regex: [/^CPNE/i] },
          { type: "required", name: "idcc", regex: [/^IDCC/i] },
        ],
      },
      {
        type: "required",
        key: "missing",
        nameMatchers: [/missing/i],
        skipRows: 1,
        columns: [],
      },
    ] as const;

    await expect(getRows(parseExcelFileStream(s, spec))).rejects.toThrow("Missing worksheets");
  });

  it("should not require optional sheets", async () => {
    const dataFixture = join(dirname(fileURLToPath(import.meta.url)), "fixtures/file.xlsx");
    const s = createReadStream(dataFixture);

    const spec = [
      {
        type: "ignore",
        nameMatchers: [/Onglet 3 - référentiel NPEC/i],
      },
      {
        type: "required",
        nameMatchers: [/Onglet 4 - CPNE-IDCC/i],
        key: "cpne-idcc",
        skipRows: 1,
        columns: [
          { type: "required", name: "cpne_code", regex: [/^Code CPNE/i] },
          { type: "required", name: "cpne_libelle", regex: [/^CPNE/i] },
          { type: "required", name: "idcc", regex: [/^IDCC/i] },
        ],
      },
      {
        type: "optional",
        key: "missing",
        nameMatchers: [/missing/i],
        skipRows: 1,
        columns: [],
      },
    ] as const;

    await expect(getRows(parseExcelFileStream(s, spec))).resolves.toBeDefined();
  });

  it("should throw if extra sheet is found", async () => {
    const dataFixture = join(dirname(fileURLToPath(import.meta.url)), "fixtures/file.xlsx");
    const s = createReadStream(dataFixture);

    const spec = [
      {
        type: "ignore",
        nameMatchers: [/Onglet 3 - référentiel NPEC/i],
      },
    ] as const;

    await expect(getRows(parseExcelFileStream(s, spec))).rejects.toThrow("Unexpected worksheet");
  });

  it("should be able to detect headers when column is set to auto", async () => {
    const dataFixture = join(dirname(fileURLToPath(import.meta.url)), "fixtures/file.xlsx");
    const s = createReadStream(dataFixture);

    const spec = [
      {
        type: "required",
        nameMatchers: [/Onglet 3 - référentiel NPEC/i],
        key: "npec",
        skipRows: 3,
        columns: "auto",
      },
      {
        type: "ignore",
        nameMatchers: [/Onglet 4 - CPNE-IDCC/i],
      },
    ] as const;

    const data = await getRows(parseExcelFileStream(s, spec));

    expect(data).toMatchSnapshot();
  });
});
