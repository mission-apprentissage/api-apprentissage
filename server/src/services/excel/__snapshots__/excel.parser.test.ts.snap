// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseExcelFileStream > should be able to detect headers when column is set to auto 1`] = `
[
  {
    "data": {
      "CPNE": "CNPE des sociétés financières",
      "Certificateur* 
(*liste non exhaustive - premier dans la fiche RNCP)": "MINISTERE DE L'EDUCATION NATIONALE ET DE LA JEUNESSE",
      "Code CPNE": 2,
      "Code RNCP": "RNCP1002",
      "Libellé de la formation": "Mise en oeuvre caoutchoucs élastomères thermoplastiques",
      "Libellé du Diplôme": "BP",
      "NPEC final": 9497,
    },
    "headers": [
      "Code RNCP",
      "Libellé de la formation",
      "Certificateur* 
(*liste non exhaustive - premier dans la fiche RNCP)",
      "Libellé du Diplôme",
      "Code CPNE",
      "CPNE",
      "NPEC final",
    ],
    "sheet": "npec",
  },
  {
    "data": {
      "CPNE": "CNPEFP des sociétés d’assistance",
      "Certificateur* 
(*liste non exhaustive - premier dans la fiche RNCP)": "MINISTERE DE L'EDUCATION NATIONALE ET DE LA JEUNESSE",
      "Code CPNE": 4,
      "Code RNCP": "RNCP1002",
      "Libellé de la formation": "Mise en oeuvre caoutchoucs élastomères thermoplastiques",
      "Libellé du Diplôme": "BP",
      "NPEC final": 9497,
    },
    "headers": [
      "Code RNCP",
      "Libellé de la formation",
      "Certificateur* 
(*liste non exhaustive - premier dans la fiche RNCP)",
      "Libellé du Diplôme",
      "Code CPNE",
      "CPNE",
      "NPEC final",
    ],
    "sheet": "npec",
  },
  {
    "data": {
      "CPNE": "CPNE du notariat",
      "Certificateur* 
(*liste non exhaustive - premier dans la fiche RNCP)": "MINISTERE DE L'EDUCATION NATIONALE ET DE LA JEUNESSE",
      "Code CPNE": 5,
      "Code RNCP": "RNCP1002",
      "Libellé de la formation": "Mise en oeuvre caoutchoucs élastomères thermoplastiques",
      "Libellé du Diplôme": "BP",
      "NPEC final": 9497,
    },
    "headers": [
      "Code RNCP",
      "Libellé de la formation",
      "Certificateur* 
(*liste non exhaustive - premier dans la fiche RNCP)",
      "Libellé du Diplôme",
      "Code CPNE",
      "CPNE",
      "NPEC final",
    ],
    "sheet": "npec",
  },
]
`;

exports[`parseExcelFileStream > should parse the excel file stream 1`] = `
[
  {
    "data": {
      "certificateur": "MINISTERE DE L'EDUCATION NATIONALE ET DE LA JEUNESSE",
      "cpne_code": 2,
      "cpne_libelle": "CNPE des sociétés financières",
      "diplome_libelle": "BP",
      "formation_libelle": "Mise en oeuvre caoutchoucs élastomères thermoplastiques",
      "npec": 9497,
      "rncp": "RNCP1002",
    },
    "headers": [
      "rncp",
      "formation_libelle",
      "certificateur",
      "diplome_libelle",
      "cpne_code",
      "cpne_libelle",
      "npec",
    ],
    "sheet": "npec",
  },
  {
    "data": {
      "certificateur": "MINISTERE DE L'EDUCATION NATIONALE ET DE LA JEUNESSE",
      "cpne_code": 4,
      "cpne_libelle": "CNPEFP des sociétés d’assistance",
      "diplome_libelle": "BP",
      "formation_libelle": "Mise en oeuvre caoutchoucs élastomères thermoplastiques",
      "npec": 9497,
      "rncp": "RNCP1002",
    },
    "headers": [
      "rncp",
      "formation_libelle",
      "certificateur",
      "diplome_libelle",
      "cpne_code",
      "cpne_libelle",
      "npec",
    ],
    "sheet": "npec",
  },
  {
    "data": {
      "certificateur": "MINISTERE DE L'EDUCATION NATIONALE ET DE LA JEUNESSE",
      "cpne_code": 5,
      "cpne_libelle": "CPNE du notariat",
      "diplome_libelle": "BP",
      "formation_libelle": "Mise en oeuvre caoutchoucs élastomères thermoplastiques",
      "npec": 9497,
      "rncp": "RNCP1002",
    },
    "headers": [
      "rncp",
      "formation_libelle",
      "certificateur",
      "diplome_libelle",
      "cpne_code",
      "cpne_libelle",
      "npec",
    ],
    "sheet": "npec",
  },
  {
    "data": {
      "cpne_code": 2,
      "cpne_libelle": "CNPE des sociétés financières",
      "idcc": 478,
    },
    "headers": [
      "cpne_code",
      "cpne_libelle",
      "idcc",
    ],
    "sheet": "cpne-idcc",
  },
  {
    "data": {
      "cpne_code": 4,
      "cpne_libelle": "CNPEFP des sociétés d’assistance",
      "idcc": 1801,
    },
    "headers": [
      "cpne_code",
      "cpne_libelle",
      "idcc",
    ],
    "sheet": "cpne-idcc",
  },
  {
    "data": {
      "cpne_code": 5,
      "cpne_libelle": "CPNE du notariat",
      "idcc": 2205,
    },
    "headers": [
      "cpne_code",
      "cpne_libelle",
      "idcc",
    ],
    "sheet": "cpne-idcc",
  },
]
`;

exports[`parseExcelFileStream > should skip parsing if spec is null 1`] = `
[
  {
    "data": {
      "cpne_code": 2,
      "cpne_libelle": "CNPE des sociétés financières",
      "idcc": 478,
    },
    "headers": [
      "cpne_code",
      "cpne_libelle",
      "idcc",
    ],
    "sheet": "cpne-idcc",
  },
  {
    "data": {
      "cpne_code": 4,
      "cpne_libelle": "CNPEFP des sociétés d’assistance",
      "idcc": 1801,
    },
    "headers": [
      "cpne_code",
      "cpne_libelle",
      "idcc",
    ],
    "sheet": "cpne-idcc",
  },
  {
    "data": {
      "cpne_code": 5,
      "cpne_libelle": "CPNE du notariat",
      "idcc": 2205,
    },
    "headers": [
      "cpne_code",
      "cpne_libelle",
      "idcc",
    ],
    "sheet": "cpne-idcc",
  },
]
`;
